# Notebook 01 — Data Collection

## Objectives
- Download the raw Telco Customer Churn dataset from Kaggle
- Inspect its structure, column types and data quality
- Save a raw baseline copy for use in subsequent notebooks

## Inputs
- Kaggle dataset: `blastchar/telco-customer-churn`
- `kaggle.json` credentials file placed in the project root

## Outputs
- `outputs/datasets/collection/WA_Fn-UseC_-Telco-Customer-Churn.csv`
- `outputs/datasets/collection/telco_churn_raw.csv`

---

## 1. Change Working Directory

In [1]:
import os

# Move up one level if we are inside jupyter_notebooks/
os.chdir('..') if os.path.basename(os.getcwd()) == 'jupyter_notebooks' else None
os.makedirs('outputs/datasets/collection', exist_ok=True)
print('Working directory:', os.getcwd())

Working directory: /workspaces/customer-churn-predictor


## 2. Install Kaggle and Set Credentials

In [ ]:
import os

# Make Kaggle look for credentials in the project root
os.environ['KAGGLE_CONFIG_DIR'] = os.getcwd()

# Set correct permissions on the credentials file (required by Kaggle)
kaggle_path = os.path.join(os.getcwd(), 'kaggle.json')
os.chmod(kaggle_path, 0o600)
print('Kaggle credentials configured.')

## 3. Download Dataset from Kaggle

In [ ]:
import kaggle

kaggle.api.authenticate()

kaggle.api.dataset_download_files(
    'blastchar/telco-customer-churn',
    path='outputs/datasets/collection/',
    unzip=True
)

print('Dataset downloaded successfully.')
print('Files:', os.listdir('outputs/datasets/collection/'))

## 4. Import Libraries

In [2]:
import pandas as pd

## 5. Load Dataset

In [3]:
df = pd.read_csv('outputs/datasets/collection/WA_Fn-UseC_-Telco-Customer-Churn.csv')
print(f'Shape: {df.shape}')
df.head()

Shape: (7043, 21)


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## 6. Data Types and Statistics

In [4]:
df.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

In [5]:
df.describe(include='all')

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
count,7043,7043,7043.000000,7043,7043,7043.000000,7043,7043,7043,7043,...,7043,7043,7043,7043,7043,7043,7043,7043.000000,7043,7043
unique,7043,2,NaN,2,2,NaN,2,3,3,3,...,3,3,3,3,3,2,4,NaN,6531,2
top,3186-AJIEK,Male,NaN,No,No,NaN,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,NaN,,No
freq,1,3555,NaN,3641,4933,NaN,6361,3390,3096,3498,...,3095,3473,2810,2785,3875,4171,2365,NaN,11,5174
mean,NaN,NaN,0.162147,NaN,NaN,32.371149,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,64.761692,NaN,NaN
std,NaN,NaN,0.368612,NaN,NaN,24.559481,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,30.090047,NaN,NaN
min,NaN,NaN,0.000000,NaN,NaN,0.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,18.250000,NaN,NaN
25%,NaN,NaN,0.000000,NaN,NaN,9.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,35.500000,NaN,NaN
50%,NaN,NaN,0.000000,NaN,NaN,29.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,70.350000,NaN,NaN
75%,NaN,NaN,0.000000,NaN,NaN,55.000000,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,89.850000,NaN,NaN


## 7. Missing Values

In [6]:
# Standard null check
print(df.isnull().sum())

# TotalCharges is object dtype — check for hidden blank strings
print('Blank strings in TotalCharges:', (df['TotalCharges'].str.strip() == '').sum())

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64
Blank strings in TotalCharges: 11


## 8. Target Distribution

In [7]:
print(df['Churn'].value_counts())
print(df['Churn'].value_counts(normalize=True).mul(100).round(1))

Churn
No     5174
Yes    1869
Name: count, dtype: int64
Churn
No     73.5
Yes    26.5
Name: proportion, dtype: float64


## 9. Save Raw Dataset

In [8]:
df.to_csv('outputs/datasets/collection/telco_churn_raw.csv', index=False)
print('Saved: telco_churn_raw.csv')

Saved: telco_churn_raw.csv


---
## Conclusions
- Dataset has 7,043 rows and 21 columns.
- `TotalCharges` is stored as object and contains 11 blank strings — to be fixed in Notebook 02.
- ~26.5% of customers churned — class imbalance will be handled with SMOTE during modelling.
- Raw CSV saved successfully.